argopy/stores/argo_index.py

```diff
@@ -0,0 +1,380 @@
+            if isinstance(self.CYC, (np.ndarray)):
+                listname = ["AllWMOs"]
+                [listname.append("CYC%0.4d" % i) for i in sorted(self.CYC)]
+            else:
+                listname = ["FULL"]
+            listname = "_".join(listname)
+        else:
+            listname = "WMO%i" % self.WMO[0]
+            if isinstance(self.CYC, (np.ndarray)):
+                listname = [listname]
+                [listname.append("CYC%0.4d" % i) for i in sorted(self.CYC)]
+                listname = "_".join(listname)
+        if len(listname) > 256:
+            listname = hashlib.sha256(listname.encode()).hexdigest()
+        return listname
+
+    def run(self, index_file):
+        """ Run search on an Argo index file
+
+        Parameters
+        ----------
+        index_file: _io.TextIOWrapper
+
+        Returns
+        -------
+        csv rows matching the request, as in-memory string. Or None.
+        """
+
+        # Define one-line search functions:
+        def search_one_wmo(index, wmo):
+            """ Search for a WMO in the argo index file
+
+            Parameters
+            ----------
+            index_file: _io.TextIOWrapper
+            wmo: int
+
+            Returns
+            -------
+            csv chunk matching the request, as a string. Or None
+            """
+            index.seek(0)
+            results = ""
+            il_read, il_loaded, il_this = 0, 0, 0
+            for line in index:
+                il_this = il_loaded
+                # if re.search("/%i/" % wmo, line.split(',')[0]):
+                if "/%i/" % wmo in line:  # much faster than re
+                    # Search for the wmo at the beginning of the file name under: /<dac>/<wmo>/profiles/
+                    results += line
+                    il_loaded += 1
+                if il_this == il_loaded and il_this > 0:
+                    break  # Since the index is sorted, once we found the float, we can stop reading the index !
+                il_read += 1
+            if il_loaded > 0:
+                return results
+            else:
+                return None
+
+        def search_any_wmo_cyc(index, cyc):
+            """ Search for a WMO in the argo index file
+
+            Parameters
+            ----------
+            index_file: _io.TextIOWrapper
+            cyc: array of integers
+
+            Returns
+            -------
+            csv chunk matching the request, as a string. Or None
+            """
+
+            def search_this(this_line):
+                # return np.any([re.search("%0.3d.nc" % c, this_line.split(',')[0]) for c in cyc])
+                return np.any(["%0.3d.nc" % c in this_line for c in cyc])
+                if np.all(cyc >= 1000):
+                    def search_this(this_line):
+                        # return np.any([re.search("%0.4d.nc" % c, this_line.split(',')[0]) for c in cyc])
+                        return np.any(["%0.4d.nc" % c in this_line for c in cyc])
```

Comment:
Unreachable "if" and function defined twice

Reply:
Yes I don't know what to do about this
The point is to make sure that we can handle float with cycle numbers larger than 1000, and for which the file name will be different, any idea ?

Reply:
The issue is that for a float with more than 1000 cycle, the cycle number 1 name will be: _0001.nc and not _001.nc

Reply:
oh shoot, now I get it !!! the return comes too early !

Reply:
yep

Reply:
Here is a fix:

```python
            if np.all(cyc >= 1000):
                def search_this(this_line):
                    # return np.any([re.search("%0.3d.nc" % c, this_line.split(',')[0]) for c in cyc])
                    return np.any(["%0.3d.nc" % c in this_line for c in cyc])
            else:
                def search_this(this_line):
                    # return np.any([re.search("%0.4d.nc" % c, this_line.split(',')[0]) for c in cyc])
                    return np.any(["%0.4d.nc" % c in this_line for c in cyc])
```